optuna/integration/botorch.py

```diff
@@ -0,0 +1,507 @@
+        sampler=SobolQMCNormalSampler(num_samples=256),
+        objective=constrained_objective,
+    )
+
+    standard_bounds = torch.zeros_like(bounds)
+    standard_bounds[1] = 1
+
+    candidates, _ = optimize_acqf(
+        acq_function=acqf,
+        bounds=standard_bounds,
+        q=1,
+        num_restarts=10,
+        raw_samples=512,
+        options={"batch_limit": 5, "maxiter": 200},
+        sequential=True,
+    )
+
+    candidates = unnormalize(candidates.detach(), bounds=bounds)
+
+    return candidates
+
+
+@experimental("2.4.0")
+def qehvi_candidates_func(
+    train_x: "torch.Tensor",
+    train_obj: "torch.Tensor",
+    train_con: Optional["torch.Tensor"],
+    bounds: "torch.Tensor",
+) -> "torch.Tensor":
+    """Expected Hypervolume Improvement (qEHVI).
+
+    The default value of ``candidates_func`` in :class:`~optuna.integration.BoTorchSampler`
+    with multi-objective optimization with or without objective constraints.
+
+    .. seealso::
+        :func:`~optuna.integration.botorch.ucb_candidates_func` for argument and return value
+        descriptions.
+    """
+
+    n_outcomes = train_obj.size(-1)
+
+    if train_con is not None:
+        train_y = torch.cat([train_obj, train_con], dim=-1)
+
+        is_feas = (train_con <= 0).all(dim=-1)
+        partitioning_y = train_obj[is_feas]
+
+        constraints = []
+        n_contraints = train_con.size(1)
+
+        for i in range(n_contraints):
+            constraints.append(lambda Z, i=i: Z[..., -n_contraints + i])
+        additional_qehvi_kwargs = {
+            "objective": IdentityMCMultiOutputObjective(outcomes=list(range(n_outcomes))),
+            "constraints": constraints,
+        }
+    else:
+        train_y = train_obj
+
+        partitioning_y = train_obj
+
+        additional_qehvi_kwargs = {}
+
+    model = SingleTaskGP(train_x, train_y, outcome_transform=Standardize(m=train_y.shape[-1]))
+    mll = ExactMarginalLogLikelihood(model.likelihood, model)
+    fit_gpytorch_model(mll)
+
+    partitioning = NondominatedPartitioning(num_outcomes=n_outcomes, Y=partitioning_y)
+
+    ref_point = train_obj.amin(dim=0) - 1e-8
+    ref_point_list = ref_point.tolist()
+
+    acqf = qExpectedHypervolumeImprovement(
+        model=model,
+        ref_point=ref_point_list,
+        partitioning=partitioning,
+        sampler=SobolQMCNormalSampler(num_samples=256),
+        **additional_qehvi_kwargs,
+    )
+
+    standard_bounds = torch.zeros_like(bounds)
+    standard_bounds[1] = 1
+
+    candidates, _ = optimize_acqf(
+        acq_function=acqf,
+        bounds=standard_bounds,
+        q=1,
+        num_restarts=20,
+        raw_samples=1024,
+        options={"batch_limit": 5, "maxiter": 200, "nonnegative": True},
+        sequential=True,
+    )
+
+    candidates = unnormalize(candidates.detach(), bounds=bounds)
+
+    return candidates
+
+
+def _get_default_candidates_func(
+    n_objectives: int, n_objective_constraints: int
```

Comment:
How about `n_objective_constraints` -> `n_constraints`?

Reply:
Removed this parameter since it was no longer necessary, now that we only condition on the number of objectives and choose between qEI and qEHVI.